### DISCUSSION TITLE
Unable to update notification recipients after upgrading to Freeform 5

### DISCUSSION THREAD

**Discussion by kate-stretton** [solspace/craft-freeform#1239] [Questions & Answers] [answered]
## Describe the bug or issue you're experiencing
When trying to update the email recipients in the Notifications tab, or add a a new recipient, the change doesn't save.

## Steps to reproduce

1. Select a form in Freeform
2. Open Notifications tab
3. Update Recipients field and click 'Save'
4. Refresh page

## Expected behavior

Able to update and save existing email recipient in Notifications tab, or add and save an additional recipient

## Craft & Plugin Info (please complete the following information):

Craft Version: 4.8.1
Freeform Version: 5.1.3
Freeform Edition: Freeform 5
Fresh Install or Upgrade: Upgrade from Freeform 4.1.13

![image](https://github.com/solspace/craft-freeform/assets/81283807/b621f1d7-a7fa-4988-84e6-2c7a23cca7ca)

**Comment by kjmartens** [comment]
Sorry for the trouble, @kate-stretton.

I have tried testing this a bit on my end and it appears to work correctly.

Are you seeing any JS errors in the browser console? 🙂

**Comment by kate-stretton** [comment]
Hello, 
Thanks for your reply. Yes, in the browser console get a 400 error (Bad Request). There's no response data.

**Comment by kjmartens** [comment]
Hi @kate-stretton,

Can you update to Freeform 5.1.8 and try it again? If the issue continues, can you provide me the Network response you're getting?

Here's how to do that (keep in mind yours won't be `options`... look for one that is in red text on your site): 🙂 

![troubleshooting-forms-network-response 19e6b0e7](https://github.com/solspace/craft-freeform/assets/1821674/d3071026-5c52-4e80-9a15-611874aa4fd2)

**Comment by kate-stretton** [comment]
Hi @kjmartens,

Thanks for your help with this! I've updated to Freeform 5.1.8 and tested updating the recipient again. Unfortunately I'm still experiencing this issue. I've checked the Network tab and there's no response data:

![image](https://github.com/solspace/craft-freeform/assets/81283807/e1d41ca9-6d42-46fa-bc5c-963e6ac1f495)

I've also now discovered that I get the same error when trying to create a new Admin notification - again a 400 error (Bad Request), and there's no response data.

**Comment by jeffreyschmitz87** [comment]
I have the same problem (version 5.1.10) and unfortunately there is no response data available

**Comment by kate-stretton** [comment]
Hi @kjmartens 
I wanted to share some additional context. I'm seeing 400 error (Bad Request) with no response data in the browser console for every change I save to the form - irrespective of whether the change is in the Layout, Notifications, Rules, Integrations or Settings tab. However, it's only saved changes in the Notifications tab that don't persist. 
I can can save changes in all the other tabs else without issue, despite seeing the 400 error in the browser console.

**Comment by kjmartens** [comment]
Sorry for the trouble @kate-stretton and @jeffreyschmitz87.

I will have one of our developers check into this and see what they can find. It's odd it just happens to a small portion of customers (I personally can't duplicate this issue either). @seandelaney may also request more information to help him debug this issue. 🙂 

I appreciate your patience!

**Comment by kjmartens** [comment]
Hi @kate-stretton and @jeffreyschmitz87,

Can you check your Craft logs and see what is showing up there? I imagine it would be located in `web-404s.log` log, but check if there's anything else logged there. 🙂

**Comment by kate-stretton** [comment]
Hi @kjmartens, 
Thank you for looking further into this. Unfortunately there's nothing showing in the web-404s.log, or any of the other logs.
The site I'm working on is a multi-site. Here's a screenshot of the Application Info, and also of the Plugins I'm using: 
![image](https://github.com/solspace/craft-freeform/assets/81283807/9cbae943-e1f8-49a5-8c6d-ba4bdd52ac2b)

I tested disabling all plugins except Freeform, but have the same issue.

**Comment by seandelaney** [comment]
@kate-stretton 

What does your Template manager settings look like in https://craft4.test/admin/freeform/settings/template-manager. Can you send me a screenshot please?

Are you using File or Database Notifications?

**Comment by seandelaney** [comment]
@kate-stretton 

> Thanks for your help with this! I've updated to Freeform 5.1.8 and tested updating the recipient again. Unfortunately I'm still experiencing this issue. I've checked the Network tab and there's no response data:

Can I please see the payload info too?

**Comment by seandelaney** [comment]
@kate-stretton 

> I wanted to share some additional context. I'm seeing 400 error (Bad Request) with no response data in the browser console for every change I save to the form - irrespective of whether the change is in the Layout, Notifications, Rules, Integrations or Settings tab.

This does not sound right. If you are getting 400 errors for any changes, then its not isolated to Notifications tab.

Did the upgrade all go okay? Any issues?

Are you able to create a new form and setup a notification?

cc: @kjmartens

**Comment by seandelaney** [comment]
@jeffreyschmitz87 

Did you also upgrade ?

Can you also please answer the same questions that I asked @kate-stretton .

Thanks


Anyways, I have setup Craft 4 with Freeform 5 (Latest release). PHP 8.2. MySQL DB though.

Not getting any errors in console or network tabs.

**Comment by seandelaney** [comment]
@kate-stretton @jeffreyschmitz87 

https://craft4.test/admin/freeform/api/forms/1/notifications

What does this request return? Can you please send me screenshots of what it sends back.

**Comment by jeffreyschmitz87** [comment]
> @jeffreyschmitz87
> 
> Did you also upgrade ?
> 
> Can you also please answer the same questions that I asked @kate-stretton .
> 
> Thanks
> 
> Anyways, I have setup Craft 4 with Freeform 5 (Latest release). PHP 8.2. MySQL DB though.
> 
> Not getting any errors in console or network tabs.

I've upgraded to the latest version (5.1.11) and unfortunately the 400 status is still there. The response from freeform/api/forms/1/notifications:

```[
  {
    "field": "5e385473-ca66-48a7-b6ec-be7ec3c3f703",
    "template": "notification.html.twig",
    "id": 31,
    "uid": "423db786-54bc-461a-926f-17d090713164",
    "enabled": true,
    "className": "Solspace\Freeform\Notifications\Types\EmailField\EmailField",
    "name": "Email Field notification 1"
  }
]
```

If I save the notification of the form with the 400 status and view the API - controller response I see an empty array as response.

Another thing I noticed is that when I open the form with the 400 status and I click on the Notifications tab and reload the page, the left side keeps 'loading' (see image below). If I then click on 'Layout' and then go back to 'Notifications', the left side does load.

![Screenshot 2024-04-12 at 09 22 19](https://github.com/solspace/craft-freeform/assets/34717885/378ee239-b347-463c-bcfb-0ec3ad8727f4)

In a certain situation it has been possible to add a notification. If I duplicate an existing form which has a notification and add a new notification to the duplicated form, it works.

Hopefully this helps!

**Comment by jeffreyschmitz87** [comment]
> @kate-stretton
> 
> > I wanted to share some additional context. I'm seeing 400 error (Bad Request) with no response data in the browser console for every change I save to the form - irrespective of whether the change is in the Layout, Notifications, Rules, Integrations or Settings tab.
> 
> This does not sound right. If you are getting 400 errors for any changes, then its not isolated to Notifications tab.
> 
> Did the upgrade all go okay? Any issues?
> 
> Are you able to create a new form and setup a notification?
> 
> cc: @kjmartens

Creating a new form and adding a new notification works!

Maybe something goes wrong if a form in Freeform 4 has no notification and after the migration to Freeform 5 a new notification can't be added?

**Comment by seandelaney** [comment]
@jeffreyschmitz87 

Thank you for the info. Very helpful!

I will test this scenario where I have a FF4 form with no notification. Upgrade to FF5 and try adding a notification and see if I can trigger this 400.

**Comment by seandelaney** [comment]
@jeffreyschmitz87 

I still cannot replicate 400 errors using a fresh FF4 installation with a form and no notifications to upgrading to FF5 and adding notifications. All worked as expected. No 400 errors.

@jeffreyschmitz87 @kate-stretton 

Next step for me is to ask for a DB backup. Please remove all sensitive info. Is this possible? email to sean@solspace.com

**Comment by kate-stretton** [comment]
Hi @seandelaney,

Thank you for looking  further into this. 

I'm using database notifications. Here's a screenshot of my Template manager settings: 
<img width="2320" alt="image" src="https://github.com/solspace/craft-freeform/assets/81283807/c17b8776-24a1-4b03-a391-ef9d62f95c21"> 

And here's one of the payload info: 
<img width="2369" alt="image" src="https://github.com/solspace/craft-freeform/assets/81283807/1d57aed6-d85f-4a4f-9521-8abe2ff60381">  

The upgrade went okay - there are no other issues. I am able to set up a notification if I create an entirely new form.

The response from freeform/api/forms/1/notifications is:
`[
  {
    "template": 1,
    "recipients": [
      {
        "email": "test@test.com",
        "name": ""
      }
    ],
    "id": 1,
    "uid": "8186be70-535d-4900-9daf-46681377167a",
    "enabled": true,
    "className": "Solspace\\Freeform\\Notifications\\Types\\Admin\\Admin",
    "name": "Admin notification"
  }
]`  

I will send you through a copy of of the database backup. Thanks!

**Comment by seandelaney** [comment]
@kate-stretton FF4 DB backup. as I want to run it through the upgrade process locally and see where notifications are going wrong. Thanks

Same for @jeffreyschmitz87 - If you can email me FF4 DB backup, that would be great,

**Comment by seandelaney** [comment]
[sean@solspace.com](mailto:sean@solspace.com)

**Comment by jeffreyschmitz87** [comment, marked as answer]
I've dug a little deeper into this issue and I think I've found the problem:

When the form notification is saved it contains error messages (https://github.com/solspace/craft-freeform/blob/v5/packages/plugin/src/Bundles/Persistence/NotificationPersistence.php#L76). These error messages are caused by a field within the form with a invalid handle name.

```
array(1) {
  ["fields"]=>
  array(1) {
    ["6c32375e-2d4d-41ca-bc2d-08554d5a7950"]=>
    array(1) {
      ["handle"]=>
      array(1) {
        [0]=>
        string(28) "Value is not a valid handle."
      }
    }
  }
}
```

One of the fields (in this case an e-mail field) contains a dash `-` inside the handle name, which causes the notification to fail when saving it. If I remove the dash from the handle name, save the form and then add a notification, this works without any problems.

Perhaps you can write a migration to remove all illegal characters from all handle names to fix this and add a check to remove illegal characters from handle names upon save? It is currently possible to add a new field with a illegal character.

**Reply by kate-stretton** [reply]
Thanks @jeffreyschmitz87. I've discovered this is the issue I've been having too. However my form field had the handle 'status' and the error message was "Value is a reserved word".

**Reply by kjmartens** [reply]
Thanks @jeffreyschmitz87 and @kate-stretton.

We'll check this out on Monday and work on a fix. 🙂

**Reply by seandelaney** [reply]
@kate-stretton @jeffreyschmitz87 

I still cannot replicate this error.

Are you saving the form within the CP when you get this error?

If your field handle has a dash in it, and you attempt to save the form, you should be getting an indicator, a little red error icon in the Layout tab to say there is a problem. Are you getting this?

![image](https://github.com/solspace/craft-freeform/assets/580403/4386cad6-afcd-4d41-b188-28ab5fddd2ad)

I just tested a FF4 installation with an email field containing a dash in it. I upgrade to FF5 fine. Migration script ran successfully. I can see the email field with the dash in its handle in the DB.

However, when I try to edit/save the form in the CP, I get that little red error icon in the layout tab. I don't get an error pointing to 
https://github.com/solspace/craft-freeform/blob/v5/packages/plugin/src/Bundles/Persistence/NotificationPersistence.php#L76. 

Also, the `NotificationPersistence` class is dependent on the form successfully saving first, to be triggered.

There must be something else in your setup causing this issue. I could really do with a DB dump to figure this out.

**Reply by jeffreyschmitz87** [reply]
@seandelaney 
I am saving the form from the CP when I get the error.

When I save the form (from the Layout tab) that has a field with a dash I get a 400 error but I see no visual indication that it failed. The strange thing is that despite the 400 error, the field is updated to the handle name with a dash.

In addition, when I switch to the Notifications tab I am trying to set up an admin - notification template and no email field notification.

Due to guidelines from our company, it is unfortunately not possible to send a database.

**Reply by seandelaney** [reply]
@jeffreyschmitz87 

> I get a 400 error but I see no visual indication that it failed. 

You should be getting this, as per the screenshot I attached y'day.

**Comment by jeffreyschmitz87** [comment]
I was a little too quick to respond! I was still running Freeform on 5.1.11. After I updated to 5.1.16 the problem was resolved. The problem has been solved for me since version 5.1.12 when updating 1 version at a time. Thank you for looking into it!

**Reply by seandelaney** [reply]
@jeffreyschmitz87 Excellent news! Thank you for sticking this out too.

**Reply by seandelaney** [reply]
@jeffreyschmitz87 Just had a look in 5.1.12... see what has changed. We switched from PUT to POST for some requests. Saving the form was one of them. Did you guys have PUT requests blocked on the server…?
